### PR TITLE
RF: auto-save datasets (part 1/X)

### PR DESCRIPTION
**Pull request by mih**

### CODE REVIEW

File: datalad/interface/save.py [hunk 1]
```diff
@@ -68,60 +97,82 @@ class Save(Interface):
...
-                # matches at least one path -> turn into submodule
-                _install_subds_inplace(
-                    ds=ds,
-                    path=utf_abspath,  # can be ignored, we don't need the return value
-                    relativepath=utf,
-                    name=None)
+        # track whether we modified anything, so it becomes
+        # possible to decide when/what to save further down
+        # and one level up
+        _modified_flag = False
```
Comment (bpoldrack) [inline datalad/interface/save.py]: Considering the next line, this line seems to be pointless. ;-)

File: datalad/interface/save.py [hunk 2]
```diff
@@ -68,60 +97,82 @@ class Save(Interface):
...
+        # now we should have a complete list of submodules to potentially
+        # recurse into
+        if recursive and (recursion_limit is None or recursion_limit):
+            # we only want immediate subdatasets, higher depths will come via
+            # recursion
+            subdss = [Dataset(opj(ds.path, subds_path))
+                      for subds_path in ds.get_subdatasets(
+                          recursive=False)]
+            # skip anything that isn't installed
+            subdss = [d for d in subdss if d.is_installed()]
```
Comment (bpoldrack) [inline datalad/interface/save.py]: `Dataset.get_subdatasets()` has parameters `fulfilled` and `absolute` ...

File: datalad/interface/save.py (hunk 1 above)
Comment (mih) [inline datalad/interface/save.py]: This deals with faulty 286 CPUs where setting a register to False need to be done twice. In-house issue...

File: datalad/interface/save.py [hunk 3]
```diff
@@ -68,60 +97,82 @@ class Save(Interface):
...
+                    # stage changes in this submodule
+                    ds.repo.add(relpath(subds.path, ds.path),
+                                git=True)
+                    _modified_flag = True
 
         if files:  # could still be none without auto add changes
             absf = [abspath(f) for f in files]
             # XXX Is there a better way to handle files in mixed repos?
             ds.repo.add(absf)
             ds.repo.add(absf, git=True)
```
Comment (bpoldrack) [inline datalad/interface/save.py]: Right now, I don't get why you add the files unconditionally. This is fine for `auto_add_changes`, but if I call `save` with `files` and `auto_add_changes=False`, should they really be added if there weren't before? Not sure about this. At least should be documented for "git ppl".

File: datalad/interface/save.py [hunk 4]
```diff
@@ -68,60 +97,82 @@ class Save(Interface):
...
+        _datalad_msg = False
+        if not message:
+            message = 'Recorded existing changes'
+            _datalad_msg = True
+
         # anything should be staged by now
-        # now however, that staged submodule changes are not considered as
+        # however, staged submodule changes are not considered as
         # `index`, hence `submodules` needs to be True too
         if ds.repo.repo.is_dirty(
```
Comment (bpoldrack) [inline datalad/interface/save.py]: If you are in the mood, please enhance `GitRepo.dirty` to pass these parameters to not directly use GitPython here. If you are not, just add a TODO.

File: datalad/interface/save.py (hunk 3 above)
Comment (bpoldrack) [inline datalad/interface/save.py]: Another thing at this point: There should be a warning or something if there are files within a subdatasets and `recursive` is `False`. Right now, there will be probably be a `CommandError` caused by `add`.

File: datalad/interface/save.py (hunk 1 above)
Comment (bpoldrack) [inline datalad/interface/save.py]: This is a joke, right? :-D

File: datalad/interface/save.py (hunk 3 above)
Comment (mih) [inline datalad/interface/save.py]: Re Error: True, fixing this right now while developing tests.

Re always add files: Files are what we are supposed to save, hence they need to be added, no?

**Comment by mih** [comment]
FTR: Weired test failure seems unrelated to the actual code changes.

File: datalad/interface/save.py (hunk 4 above)
Comment (mih) [inline datalad/interface/save.py]: On record now: #781

**Comment by mih** [comment]
Will bounce this to get another CI run.

File: datalad/interface/save.py (hunk 1 above)
Comment (mih) [inline datalad/interface/save.py]: I am never joking ;-)

(this line is the anchor for the comment, nothing more)

**Comment by mih** [comment]
All pass! HIT THAT BUTTON!

File: datalad/interface/save.py (hunk 3 above)
Comment (bpoldrack) [inline datalad/interface/save.py]: Sure, but we have an `add` command. Commiting unstaged files without `-a` might at least be confusing to git ppl. While I do see, that a datalad user may expect this behaviour and it is indeed somewhat reasonable, if the files were explicitly passed, it should be mentioned in the docs for "power-users" (?).

File: datalad/interface/save.py [hunk 5]
```diff
@@ -68,61 +98,105 @@ class Save(Interface):
...
+                None if recursion_limit is None else max(recursion_limit - 1, 0)
+            for subds in subdss:
+                subds_modified = Save.__call__(
+                    message=message,
+                    files=files,
+                    dataset=subds,
+                    auto_add_changes=auto_add_changes,
+                    version_tag=version_tag,
+                    recursive=recursive and (prop_recursion_limit is None or prop_recursion_limit > 0),
+                    recursion_limit=None if recursion_limit is None else max(recursion_limit - 1, 0),
```
Comment (bpoldrack) [inline datalad/interface/save.py]: Seems to be line 158. So: `recursion_limit=prop_recursion_limit` ?

File: datalad/interface/save.py [hunk 6]
```diff
@@ -68,61 +98,105 @@ class Save(Interface):
...
+                    p, recursion_limit=1) for p in files]
+            # skip anything that isn't installed, or this dataset
+            subdss = [d for d in subdss if d.is_installed() and d != ds]
+
+            prop_recursion_limit = \
+                None if recursion_limit is None else max(recursion_limit - 1, 0)
+            for subds in subdss:
+                subds_modified = Save.__call__(
+                    message=message,
+                    files=files,
```
Comment (bpoldrack) [inline datalad/interface/save.py]: Line 163 + line 177ff:

Why not filtering files first and pass only fitting ones to recursive calls? Right now the very same paths are checked over and over again.

File: datalad/interface/save.py (hunk 5 above)
Comment (mih) [inline datalad/interface/save.py]: Good point! Will fix.

File: datalad/interface/save.py (hunk 6 above)
Comment (mih) [inline datalad/interface/save.py]: Ack, will try that.

File: datalad/interface/save.py (hunk 3 above)
Comment (mih) [inline datalad/interface/save.py]: True, but as this doesn't change the current behavior, I defer this to #784

**Pull request merged**